src/libraries/System.Net.Primitives/src/System/Net/CookieContainer.cs

```diff
@@ -586,9 +587,9 @@ public void Add(CookieCollection cookies)
             {
                 throw new ArgumentNullException(nameof(cookies));
             }
-            foreach (Cookie c in cookies)
+            foreach (Cookie? c in cookies)
```

Comment:
Why is this ? needed?  CookieCollection's enumerator should be defined to return `Cookie` rather than `Cookie?`, no?

Reply:
Compiler writes error. I guess because CookieCollection is defined as ICollection.

Reply:
Getting the same error. Only if I explicitly cast the collection does it go away:
```csharp
foreach (Cookie c in (ICollection<Cookie>)cookies)
```
but arguable this makes things worse. I wonder why it resolves to the ICollection by default.

Reply:
> arguably this makes things worse

Seems fine to me.

> wonder why it resolves to ICollection by default

Because the public GetEnumerator method returns the non-generic interface.

Reply:
@jcouv, any better way to work around this in the CookieCollection implementation/ref, that wouldn't be a breaking change?

Reply:
@stephentoub I don't see a better option without breaking compat. Using `foreach (Cookie? in ..)` doesn't cause much harm here.

Reply:
> Using foreach (Cookie? in ..) doesn't cause much harm here.

Really?  It means anyone enumerating cookies is going to think they can be null and need to be checked for null.

Reply:
Oh, I see. Your concern is with other users of this API, not just this specific instance. Sadly, no, I don't have a better idea. Maybe worth starting a thread with LDM in case someone has an idea.

Reply:
> Oh, I see. Your concern is with other users of this API, not just this specific instance.

Right.

One thought would be, for the purposes of nullability, seeing whether the type being foreach'd implements the generic `IEnumerable<T>`, and using the nullability of that `T` even if it binds to the non-generic `IEnumerable`... is that within the realm of feasible/reasonable?

Reply:
Nice idea. Hard-coding knowledge of `IEnumerable` and `IEnumerable<T>` would work.
Let's bring it up. I think we're pretty amenable to such rules in nullable analysis, since it is heuristics-based. I suspect the implementation could get messy, but we'll only know by trying.

Reply:
https://github.com/dotnet/csharplang/issues/3214

Reply:
So what is the resolution for the current PR? Should we leave as-is?

Reply:
> One thought would be, for the purposes of nullability, seeing whether the type being foreach'd implements the generic IEnumerable<T>, and using the nullability of that T even if it binds to the non-generic IEnumerable, if the iteration variable has the same type as the generic... is that within the realm of feasible/reasonable?

Isn't it the case that we can create pathological implementations that return null values on the untyped enumerators only?

Reply:
Since @stephentoub fine with `(ICollection<Cookie>)` I can pull the change.

Reply:
> Isn't it the case that we can create pathological implementations that return null values on the untyped enumerators only?

Sure. Implementations can also say they don't return null and still do so. The system is fallible.